### DISCUSSION TITLE
How to translate makefile to b2 jamfile conditional

### DISCUSSION THREAD

**Discussion by fasxmut** [bfgroup/b2#203] [Q&A] [answered]
How to translate the following makefile code to b2 jamfile conditional requirement:

```
ifdef ABC
    CXXFLAGS += -O2
endif
```

**Comment by grisumbras** [comment]
What's ABC in this case? An environment variable?

**Reply by fasxmut** [reply]
> What's ABC in this case? An environment variable?

I want to translate this:
```
ifdef PROFILE
CPPFLAGS += -pg
endif
```

But then I found b2 has already processed it:

```
b2 -q -n variant=profile > log.log
```

Then flag -pg can be found in log.log .

**Comment by grisumbras** [comment]
Yes, it's one of the possible values for `variant` feature. In general, using features is preferred to manually adding flags. The list of built-in features and their values can be seen here: https://www.boost.org/doc/libs/1_80_0/tools/build/doc/html/index.html#bbv2.overview.builtins.features.

**Comment by grisumbras** [comment, marked as answer]
Returning to your original question, if you do need something like that you can go the most primitive route and add a similar construct to `project-config.jam`:

```
import os ;
local ABC = [ os.environ ABC ] ;
if $(ABC) {
    ABC = <cxxflags>-XYZ ;
}
# project-config is a parent of root project and its properties are applied to it
project :  default-build $(ABC) ;
```
With a POSIX shell this can be used as
```shell
ABC=1 b2
```

You can also pass that value via a b2 CLI flag (or support both):
```
import os ;
ABC ?= [ os.environ ABC ] ; # only replaces the variable's value if it's empty
if $(ABC) {
    ABC = <cxxflags>-XYZ ;
}
project :  default-build $(ABC) ;
```

And invoke b2 like this:
```shell
b2 -s ABC=1
```

And you can go fancier with `option` module:
```
import option ;
local ABC = [ option.get ABC : : <cxxflags>-XYZ ] ;
project :  default-build $(ABC) ;
```

And invoke b2 like this:
```shell
b2 --ABC # NOT ABC=1
```

Finally, you can do it by creating a b2 feature. In this case, a composite feature will do:
```
import feature ;
feature.feature ABC : off on : composite link-incompatible ;
feature.compose <ABC>on : <cxxflags>-XYZ ;
```

The benefit is that now you can ask for 2 builds at the same time, with ABC enabled and disabled:
```shell
b2 ABC=on,off
```

Some docs on features: https://www.boost.org/doc/libs/1_80_0/tools/build/doc/html/index.html#bbv2.extending.features

**Reply by fasxmut** [reply]
These are very useful, thank you.
